File: 01-api-pedidos/manual/bun/test/api.test.ts
import { test, expect, beforeAll, afterAll } from "bun:test";
import { app } from "../src/index";

const BASE_URL = "http://localhost:3000";

beforeAll(async () => {
  await new Promise((resolve) => setTimeout(resolve, 100));
});

afterAll(() => {
  app.stop();
});

test("testCrudProduto", async () => {
  // CREATE
  const createRes = await fetch(`${BASE_URL}/api/produtos`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      nome: "Produto Teste CRUD",
      descricao: "Descricao do produto de teste",
      preco: 99.90,
      estoque: 10,
      categoria: "Testes",
    }),
  });
  expect(createRes.status).toBe(201);
  const created = (await createRes.json()) as { id: number; nome: string; preco: number };
  expect(created).toHaveProperty("id");
  expect(created.nome).toBe("Produto Teste CRUD");
  expect(created.preco).toBe(99.90);
  const prodId = created.id;

  // READ
  const getRes = await fetch(`${BASE_URL}/api/produtos/${prodId}`);
  expect(getRes.status).toBe(200);
  const fetched = (await getRes.json()) as { id: number; nome: string };
  expect(fetched.id).toBe(prodId);
  expect(fetched.nome).toBe("Produto Teste CRUD");

  // UPDATE
  const updateRes = await fetch(`${BASE_URL}/api/produtos/${prodId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      nome: "Produto Atualizado",
      preco: 149.90,
      estoque: 20,
    }),
  });
  expect(updateRes.status).toBe(200);
  const updated = (await updateRes.json()) as { nome: string; preco: number };
  expect(updated.nome).toBe("Produto Atualizado");
  expect(updated.preco).toBe(149.90);

  // DELETE
  const deleteRes = await fetch(`${BASE_URL}/api/produtos/${prodId}`, {
    method: "DELETE",
  });
  expect(deleteRes.status).toBe(204);

  // VERIFY DELETED
  const goneRes = await fetch(`${BASE_URL}/api/produtos/${prodId}`);
  expect(goneRes.status).toBe(404);
});

test("testBuscaCaseInsensitive", async () => {
  const res = await fetch(`${BASE_URL}/api/produtos?busca=notebook`);
  expect(res.status).toBe(200);

  const body = (await res.json()) as { data: { nome: string }[]; total: number };

  expect(body.total).toBeGreaterThan(0);
});

test("testCriarPedido", async () => {
  const res = await fetch(`${BASE_URL}/api/pedidos`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cliente_id: 1,
      itens: [
        { produto_id: 1, quantidade: 2 },
        { produto_id: 2, quantidade: 1 },
      ],
    }),
  });

  expect(res.status).toBe(201);
  const pedido = (await res.json()) as {
    id: number;
    cliente_id: number;
    status: string;
    total: number;
  };
  expect(pedido).toHaveProperty("id");
  expect(pedido.cliente_id).toBe(1);
  expect(pedido.status).toBe("pendente");
  expect(pedido.total).toBeGreaterThan(0);

  // Verify details via GET
  const detalheRes = await fetch(`${BASE_URL}/api/pedidos/${pedido.id}`);
  expect(detalheRes.status).toBe(200);
  const detalhe = (await detalheRes.json()) as {
    id: number;
    itens: unknown[];
    cliente_nome: string;
  };
  expect(detalhe.id).toBe(pedido.id);
  expect(Array.isArray(detalhe.itens)).toBe(true);
  expect(detalhe).toHaveProperty("cliente_nome");
});

test("testPrecoNegativoRejeitado", async () => {
  const res = await fetch(`${BASE_URL}/api/produtos`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      nome: "Produto Preco Negativo",
      preco: -10.00,
      estoque: 5,
      categoria: "Teste",
    }),
  });

  expect(res.status).toBe(400);
});

test("testListarClientes", async () => {
  const res = await fetch(`${BASE_URL}/api/clientes`);
  expect(res.status).toBe(200);

  const clientes = (await res.json()) as { id: number; nome: string; email: string }[];
  expect(Array.isArray(clientes)).toBe(true);
  expect(clientes.length).toBe(20);

  // Detail with order history
  const detalheRes = await fetch(`${BASE_URL}/api/clientes/1`);
  expect(detalheRes.status).toBe(200);
  const cliente = (await detalheRes.json()) as {
    pedidos: unknown[];
    total_pedidos: number;
    total_gasto: number;
  };
  expect(cliente).toHaveProperty("pedidos");
  expect(cliente).toHaveProperty("total_pedidos");
  expect(cliente).toHaveProperty("total_gasto");
});

test("testPaginacaoTotalPages", async () => {
  const res = await fetch(`${BASE_URL}/api/produtos?page=1&limit=7`);
  expect(res.status).toBe(200);

  const body = (await res.json()) as { total: number; limit: number; totalPages: number };
  const expectedTotalPages = Math.ceil(body.total / body.limit);

  expect(body.totalPages).toBe(expectedTotalPages);
});

test("testDashboardResumo", async () => {
  const res = await fetch(`${BASE_URL}/api/dashboard/resumo`);
  expect(res.status).toBe(200);

  const body = (await res.json()) as {
    total_vendas: number;
    total_pedidos: number;
    ticket_medio: number;
  };

  expect(body).toHaveProperty("total_vendas");
  expect(body).toHaveProperty("total_pedidos");
  expect(body).toHaveProperty("ticket_medio");

  expect(typeof body.total_vendas).toBe("number");
  expect(typeof body.total_pedidos).toBe("number");
  expect(typeof body.ticket_medio).toBe("number");

  expect(body.total_vendas).toBeGreaterThan(0);
  expect(body.total_pedidos).toBeGreaterThan(0);
  expect(body.ticket_medio).toBeGreaterThan(0);
});

test("testTotalPedidoComQuantidade", async () => {
  const createProdRes = await fetch(`${BASE_URL}/api/produtos`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      nome: "Produto Para Teste Quantidade",
      descricao: "Produto de teste",
      preco: 100.00,
      estoque: 50,
      categoria: "Teste",
    }),
  });
  expect(createProdRes.status).toBe(201);
  const prodBody = (await createProdRes.json()) as { id: number };
  const prodId = prodBody.id;

  const res = await fetch(`${BASE_URL}/api/pedidos`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      cliente_id: 1,
      itens: [{ produto_id: prodId, quantidade: 3 }],
    }),
  });

  expect(res.status).toBe(201);
  const pedido = (await res.json()) as { id: number; total: number };

  expect(pedido.total).toBe(300.00);
});

test("testFiltroStatus", async () => {
  const resAll = await fetch(`${BASE_URL}/api/pedidos?limit=200`);
  expect(resAll.status).toBe(200);
  const allBody = (await resAll.json()) as { data: { status: string }[]; total: number };
  const totalGeral = allBody.total;

  const res = await fetch(`${BASE_URL}/api/pedidos?status=entregue&limit=200`);
  expect(res.status).toBe(200);
  const body = (await res.json()) as { data: { status: string }[]; total: number };

  const allEntregue = body.data.every((p) => p.status === "entregue");

  expect(allEntregue).toBe(true);
  expect(body.total).toBeLessThan(totalGeral);
});

File: 01-api-pedidos/ai/kotlin/src/test/kotlin/com/treinamento/ApiTest.kt
package com.treinamento

import com.treinamento.models.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.server.testing.*
import kotlinx.serialization.json.*
import kotlin.test.*

class ApiTest {

    private fun buildClient(builder: ApplicationTestBuilder) =
        builder.createClient {
            install(ContentNegotiation) {
                json(kotlinx.serialization.json.Json {
                    ignoreUnknownKeys = true
                    isLenient = true
                    namingStrategy = JsonNamingStrategy.SnakeCase
                })
            }
        }

    @Test
    fun testCrudProduto() = testApplication {
        application { module() }
        val client = buildClient(this)

        // CREATE
        val createResp = client.post("/api/produtos") {
            contentType(ContentType.Application.Json)
            setBody(
                ProdutoRequest(
                    nome      = "Produto Teste CRUD",
                    descricao = "Descrição do produto de teste",
                    preco     = 99.90,
                    estoque   = 10,
                    categoria = "Testes"
                )
            )
        }
        assertEquals(HttpStatusCode.Created, createResp.status)
        val created = createResp.body<Produto>()
        assertEquals("Produto Teste CRUD", created.nome)
        assertEquals(99.90, created.preco)
        val prodId = created.id

        // READ
        val getResp = client.get("/api/produtos/$prodId")
        assertEquals(HttpStatusCode.OK, getResp.status)
        val fetched = getResp.body<Produto>()
        assertEquals(prodId, fetched.id)
        assertEquals("Produto Teste CRUD", fetched.nome)

        // UPDATE
        val updateResp = client.put("/api/produtos/$prodId") {
            contentType(ContentType.Application.Json)
            setBody(
                ProdutoRequest(
                    nome      = "Produto Atualizado",
                    descricao = "Descrição atualizada",
                    preco     = 149.90,
                    estoque   = 20,
                    categoria = "Testes"
                )
            )
        }
        assertEquals(HttpStatusCode.OK, updateResp.status)
        val updated = updateResp.body<Produto>()
        assertEquals("Produto Atualizado", updated.nome)
        assertEquals(149.90, updated.preco)

        // DELETE
        val deleteResp = client.delete("/api/produtos/$prodId")
        assertEquals(HttpStatusCode.NoContent, deleteResp.status)

        // Confirm gone
        val goneResp = client.get("/api/produtos/$prodId")
        assertEquals(HttpStatusCode.NotFound, goneResp.status)
    }

    @Test
    fun testBuscaCaseInsensitive() = testApplication {
        application { module() }
        val client = buildClient(this)

        val resp = client.get("/api/produtos?busca=notebook")
        assertEquals(HttpStatusCode.OK, resp.status)
        val page = resp.body<PaginatedResponse<Produto>>()

        assertTrue(
            page.total > 0,
            "Busca por 'notebook' deveria encontrar 'Notebook Pro'"
        )
    }

    @Test
    fun testCriarPedido() = testApplication {
        application { module() }
        val client = buildClient(this)

        val pedidoResp = client.post("/api/pedidos") {
            contentType(ContentType.Application.Json)
            setBody(
                CriarPedidoRequest(
                    clienteId = 1,
                    itens     = listOf(
                        ItemRequest(produtoId = 1, quantidade = 2),
                        ItemRequest(produtoId = 2, quantidade = 1)
                    )
                )
            )
        }
        assertEquals(HttpStatusCode.Created, pedidoResp.status)
        val pedido = pedidoResp.body<Pedido>()
        assertEquals(1, pedido.clienteId)
        assertEquals("pendente", pedido.status)
        assertTrue(pedido.total > 0, "Total deve ser maior que zero")

        // Verify details via GET
        val detalheResp = client.get("/api/pedidos/${pedido.id}")
        assertEquals(HttpStatusCode.OK, detalheResp.status)
        val detalhe = detalheResp.body<PedidoDetalhe>()
        assertEquals(pedido.id, detalhe.id)
        assertTrue(detalhe.itens.isNotEmpty())
        assertTrue(detalhe.clienteNome.isNotBlank())
    }

    @Test
    fun testPrecoNegativoRejeitado() = testApplication {
        application { module() }
        val client = buildClient(this)

        val resp = client.post("/api/produtos") {
            contentType(ContentType.Application.Json)
            setBody(
                ProdutoRequest(
                    nome      = "Produto Preco Negativo",
                    descricao = "Produto com preco invalido",
                    preco     = -10.00,
                    estoque   = 5,
                    categoria = "Teste"
                )
            )
        }
        assertEquals(
            HttpStatusCode.BadRequest,
            resp.status,
            "API deve rejeitar produto com preco negativo"
        )
    }

    @Test
    fun testListarClientes() = testApplication {
        application { module() }
        val client = buildClient(this)

        val resp = client.get("/api/clientes")
        assertEquals(HttpStatusCode.OK, resp.status)
        val clientes = resp.body<List<Cliente>>()
        assertEquals(20, clientes.size)

        // Detail with order history
        val detalheResp = client.get("/api/clientes/1")
        assertEquals(HttpStatusCode.OK, detalheResp.status)
        val detalhe = detalheResp.body<ClienteDetalhe>()
        assertTrue(detalhe.pedidos.isNotEmpty(), "Cliente deve ter pedidos")
        assertTrue(detalhe.totalPedidos > 0, "Cliente deve ter total_pedidos > 0")
        assertTrue(detalhe.totalGasto > 0, "Cliente deve ter total_gasto > 0")
    }

    @Test
    fun testPaginacaoTotalPages() = testApplication {
        application { module() }
        val client = buildClient(this)

        val resp = client.get("/api/produtos?page=1&limit=7")
        assertEquals(HttpStatusCode.OK, resp.status)
        val page = resp.body<PaginatedResponse<Produto>>()

        val expectedTotalPages = kotlin.math.ceil(page.total.toDouble() / page.limit).toInt()

        assertEquals(
            expectedTotalPages,
            page.totalPages,
            "totalPages deveria ser ceil(${page.total}/${page.limit})=$expectedTotalPages, mas foi ${page.totalPages}"
        )
    }

    @Test
    fun testDashboardResumo() = testApplication {
        application { module() }
        val client = buildClient(this)

        val resp = client.get("/api/dashboard/resumo")
        assertEquals(HttpStatusCode.OK, resp.status)
        val resumo = resp.body<DashboardResumo>()
        assertTrue(resumo.totalVendas > 0, "total_vendas deve ser positivo")
        assertTrue(resumo.totalPedidos > 0, "total_pedidos deve ser positivo")
        assertTrue(resumo.ticketMedio > 0, "ticket_medio deve ser positivo")
    }

    @Test
    fun testTotalPedidoComQuantidade() = testApplication {
        application { module() }
        val client = buildClient(this)

        val createResp = client.post("/api/produtos") {
            contentType(ContentType.Application.Json)
            setBody(
                ProdutoRequest(
                    nome      = "Produto Para Teste Quantidade",
                    descricao = "Produto de teste",
                    preco     = 100.00,
                    estoque   = 50,
                    categoria = "Teste"
                )
            )
        }
        assertEquals(HttpStatusCode.Created, createResp.status)
        val produto = createResp.body<Produto>()

        val quantidade = 3
        val pedidoResp = client.post("/api/pedidos") {
            contentType(ContentType.Application.Json)
            setBody(
                CriarPedidoRequest(
                    clienteId = 1,
                    itens     = listOf(ItemRequest(produtoId = produto.id, quantidade = quantidade))
                )
            )
        }
        assertEquals(HttpStatusCode.Created, pedidoResp.status)
        val pedido = pedidoResp.body<Pedido>()

        assertEquals(
            300.00,
            pedido.total,
            0.01,
            "Total do pedido deveria ser 300.00 (3 x 100.00), mas foi ${pedido.total}"
        )
    }

    @Test
    fun testFiltroStatus() = testApplication {
        application { module() }
        val client = buildClient(this)

        val allResp = client.get("/api/pedidos?limit=200")
        assertEquals(HttpStatusCode.OK, allResp.status)
        val allPedidos = allResp.body<PaginatedResponse<Pedido>>()
        val totalGeral = allPedidos.total

        val filteredResp = client.get("/api/pedidos?status=entregue&limit=200")
        assertEquals(HttpStatusCode.OK, filteredResp.status)
        val filteredPedidos = filteredResp.body<PaginatedResponse<Pedido>>()

        assertTrue(
            filteredPedidos.total < totalGeral,
            "Filtro por status deve retornar menos pedidos que o total"
        )
        assertTrue(
            filteredPedidos.data.all { it.status == "entregue" },
            "Todos os pedidos retornados devem ter status 'entregue'"
        )
    }
}
